packages/dev/core/src/Maths/tensor.ts

```diff
@@ -49,15 +59,15 @@ export interface Tensor<V extends unknown[] = unknown[]> {
      * @param index defines the offset in source array
      * @returns the current instance
      */
-    toArray(array: FloatArray, index?: number): this;
+    toArray(array: ArrayLike<TensorValueElement<V>>, index?: number): this;

```

Comment:
So, just wondering - V extends unknown[] - meaning it is an array. What you are doing is inferring the value of this array to, in general, generate an array (or ArrayLike) of this value, which is exactly V. (of course, types only)
I get it that ArrayLike does not exactly mean Array, I just don't quite get what should be the difference between V and `ArrayLike<TensorValueElement<V>>`, which feels very over-engineered and hard to follow.

Also, in what case is it not a number? We don't have a class using anything other than a number, and I don't see us supporting anything else, mainly because of the final consumption of these types.

Reply:
V is the value of the Tensor. It is not just an array. For example in `Matrix`, `V` is a `Tuple<Tuple<number, 4>, 4>`, which "decays" to `number[][]`.

Since `FromArray` expects an array of floats, passing in `V` does not make sense.

Reply:
Not 100% what I meant, but I guess it is a bit hard to explain that.

Eventually, these are all numbers. Wether number[] or Float32Array. Nothing more. Seeing types like this:

```javascript
export type Flatten<A extends unknown[]> = IsTuple<A> extends true ? FlattenTuple<A> : FlattenArray<A>;

type _Tuple<T, N extends number, R extends unknown[] = Empty> = R["length"] extends N ? R : _Tuple<T, N, [T, ...R]>;

/**
 * Creates a tuple of T with length N
 */
export type Tuple<T, N extends number> = _Tuple<T, N>;
```

ot this:

```javascript
/**
 * Extracts the element type of a Tensor value
 */
export type TensorValueElement<V extends unknown[]> = Flatten<V> extends Array<infer E> ? E : never;

/**
 * Extracts the element type of a Tensor
 */
export type TensorElement<T> = T extends Tensor<infer V> ? TensorValueElement<V> : never;
```

Feels like an over-complication of a simple process. They are also quite hard to read and understand. Half of them are already in, but I am not 100% happy with adding more.

What was wrong with a FloatArray? what can be the type of the element of V, that will break this?

Reply:
In all of the BJS use cases, `V` is a tuple of numbers.

If used in `FromArray`, then the array passed to the function would need an exact number of elements.

That would mean loading multiple `Vector3`s from a single array, which had the data (i.e. a single array with all the vertices) would not be possible since `Vector3.FromArray` would expect `[number, number, number]`.

In `Matrix`, if `V` is used directly as the type of `array` in `FromArray`, then it would be `Tuple<Tuple<number, 4>, 4>` which expands to `[
[number, number, number, number],
[number, number, number, number],
[number, number, number, number],
[number, number, number, number]
]`,

This is obviously not an array and the reason why the value can't be used as the type of the array. The value type is intended to be as narrow as possible, while `FromArray` is supposed to take a flat array of the elements of the value.

Without using tuples, much of the Tensor API would be more difficult to work with, since the base class would use arrays for everything that required floats (so it would not resolve properly), like in `addInPlaceFromFloats`. This would give an error with existing implementations and would mean that `addInPlaceFromFloats` implementations would have to use a rest parameter, casting, array options, and more complex types to achieve the same result.

Reply:
What was wrong with a FloatArray? Where does it break? Sorry if you answered that already, but having this as a public type `DeepImmutable<ArrayLike<TensorValueElement<V>>>` is quite unreadable for people unfamiliar with the framework or the terms.

Reply:
`FloatArray` is not compatible with `V` since `V extends unknown[]`. Unfortunately, we can not have `V extends number[]` since that would break `Matrix`.

Since `ArrayLike` is already `readonly` I can remove the `DeepImmutable`, which should clean up the types more.

What do you think about renaming `TensorValueElement` to `TensorElement`?

Reply:
that's not going to change the fact that from a single type we moved to a 4-level generic object.

FloatArray is fully compatible with matrix right now, we are also talking about the toArray and fromArray - those functions take an array as their first property. It might be a deep immutable array-like with element type of the original generic type that was used to create it, but isn't it simpler to use FloatArray? Where does it break? in what case is it not compatible?

Reply:
`TensorElement` does not work since it gives an error on `VectorStatic`: (Type instantiation is recursive and possibly infinite)

Reply:
> that's not going to change the fact that from a single type we moved to a 4-level generic object.
>
> FloatArray is fully compatible with matrix right now, we are also talking about the toArray and fromArray - those functions take an array as their first property. It might be a deep immutable array-like with element type of the original generic type that was used to create it, but isn't it simpler to use FloatArray? Where does it break? in what case is it not compatible?

I had this exact thought. After messing around with the types for a while, I found that it is possible to just define a type `TensorValue = number[] | TensorValue[]`, which means in `Tensor<V>`, `V` is guaranteed to be an n-depth array of number. 

I was able to confirm using a debug type called [`ExpandRecursively`](https://stackoverflow.com/a/57683652/17637456), that `TensorValueElement<TensorValue>` is `number`, which means we can do away with all the confusing types related to Tensor value elements.

I did have to make a minor change to `FlattenArray` that allows it to handle recursive arrays.

Reply:
Also, I've updated the relevant functions to use `FloatArray`